Nim doesn't catch wrong var {.global.} initialization

Test

``` Nim
proc foo =
  let a = 0
  var b {.global.} = a # Nim should complain about this
foo()
```

Result:

``` Nim
>nim c test
[...]\nimcache\compiler_test.c:100:12: error: 'a' undeclared (first use in this f
unction)
  b_97006 = a;
            ^
Error: execution of an external program failed: [...]
```

---

``` Nim
>nim -v
Nim Compiler Version 0.12.0 (2015-10-28) [Windows: i386]
Copyright (c) 2006-2015 by Andreas Rumpf

git hash: 22a702868bfd242ff76bf86ea00a278f979a112e
active boot switches: -d:release
```

> this works in current devel https://github.com/nim-lang/Nim/commit/3b973f3fc8a87a42b2b3855a56d87f6646fe76ac, and similar issues are closed, guess it's just reopened by adding a label.

It doesn't work if the proc gets called

```nim
proc foo =
  let a = 0
  var b {.global.} = a # Nim should complain about this

foo()
```
```
C:\Users\blue\Documents\GitHub\Nim\test3.nim(3, 7) Hint: 'b' is declared but not used [XDeclaredButNotUsed]
CC: ../../../.choosenim/toolchains/nim-#devel/lib/system/exceptions.nim
CC: ../../../.choosenim/toolchains/nim-#devel/lib/system.nim
CC: test3.nim
C:\Users\blue\nimcache\test3_d\@mtest3.nim.c: In function 'NimMainModule':
C:\Users\blue\nimcache\test3_d\@mtest3.nim.c:229:23: error: 'a' undeclared (first use in this function)
  229 |         b__test51_3 = a;
      |                       ^
```

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -591,6 +591,25 @@ proc msgSymChoiceUseQualifier(c: PContext; n: PNode; note = errGenerated) =
     inc i
   message(c.config, n.info, note, err)
 
+proc isUseLocalVar(n: PNode): bool =
+  for z in 1 ..< n.len:
+    if n[z].kind == nkSym and n[z].sym.kind in {skVar, skLet} and 
+      not ({sfGlobal, sfPure} <= n[z].sym.flags):
+      return true
+    elif n[z].kind == nkCall:
+      if isUseLocalVar(n[z]):
+        return true
+
+proc globalVarInitCheck(c: PContext, s: PSym, n: PNode) =
+  let isGlobal = {sfGlobal, sfPure} <= s.flags
+  if isGlobal and n.kind == nkCall and isUseLocalVar(n):
+    localError(c.config, n.info, errCannotAsignToGlobal)
+  if isGlobal and n.kind == nkSym and 
+      n.sym.kind in {skVar, skLet} and not 
+      ({sfGlobal, sfPure} <= n.sym.flags or
+        sfCompileTime in n.sym.flags):
+    localError(c.config, n.info, errCannotAsignToGlobal)
+
 proc semVarOrLet(c: PContext, n: PNode, symkind: TSymKind): PNode =
   var b: PNode
   result = copyNode(n)
@@ -741,17 +760,7 @@ proc semVarOrLet(c: PContext, n: PNode, symkind: TSymKind): PNode =
         vm.setupCompileTimeVar(c.module, c.idgen, c.graph, x)
       if v.flags * {sfGlobal, sfThread} == {sfGlobal}:
         message(c.config, v.info, hintGlobalVar)
-      let isGlobal = {sfGlobal, sfPure} <= v.flags
-      if isGlobal and def.kind == nkCall:
-        for z in 1 ..< def.len:
-          if def[z].kind == nkSym and def[z].sym.kind in {skVar, skLet} and 
-            not ({sfGlobal, sfPure} <= def[z].sym.flags):
-            localError(c.config, def.info, errCannotAsignToGlobal)
-      if isGlobal and 
-          def.kind == nkSym and def.sym.kind in {skVar, skLet} and not 
-          ({sfGlobal, sfPure} <= def.sym.flags or
-            sfCompileTime in def.sym.flags):
-        localError(c.config, def.info, errCannotAsignToGlobal)
+      globalVarInitCheck(c, v, def)
       suggestSym(c.graph, v.info, v, c.graph.usageSym)
 
 proc semConst(c: PContext, n: PNode): PNode =

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -36,6 +36,7 @@ const
   errRecursiveDependencyX = "recursive dependency: '$1'"
   errRecursiveDependencyIteratorX = "recursion is not supported in iterators: '$1'"
   errPragmaOnlyInHeaderOfProcX = "pragmas are only allowed in the header of a proc; redefinition of $1"
+  errCannotAsignToGlobal = "cannot asign to global variable"
 
 proc semDiscard(c: PContext, n: PNode): PNode =
   result = n
@@ -740,11 +741,17 @@ proc semVarOrLet(c: PContext, n: PNode, symkind: TSymKind): PNode =
         vm.setupCompileTimeVar(c.module, c.idgen, c.graph, x)
       if v.flags * {sfGlobal, sfThread} == {sfGlobal}:
         message(c.config, v.info, hintGlobalVar)
-      if {sfGlobal, sfPure} <= v.flags and 
-          def.kind == nkSym and def.sym.kind in {skVar, skLet, skParam} and not 
+      let isGlobal = {sfGlobal, sfPure} <= v.flags
+      if isGlobal and def.kind == nkCall:
+        for z in 1 ..< def.len:
+          if def[z].kind == nkSym and def[z].sym.kind in {skVar, skLet} and 
+            not ({sfGlobal, sfPure} <= def[z].sym.flags):
+            localError(c.config, def.info, errCannotAsignToGlobal)
+      if isGlobal and 
+          def.kind == nkSym and def.sym.kind in {skVar, skLet} and not 
           ({sfGlobal, sfPure} <= def.sym.flags or
             sfCompileTime in def.sym.flags):
-        localError(c.config, def.info, "cannot asign to global variable")
+        localError(c.config, def.info, errCannotAsignToGlobal)
       suggestSym(c.graph, v.info, v, c.graph.usageSym)
 
 proc semConst(c: PContext, n: PNode): PNode =

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -740,7 +740,11 @@ proc semVarOrLet(c: PContext, n: PNode, symkind: TSymKind): PNode =
         vm.setupCompileTimeVar(c.module, c.idgen, c.graph, x)
       if v.flags * {sfGlobal, sfThread} == {sfGlobal}:
         message(c.config, v.info, hintGlobalVar)
-
+      if {sfGlobal, sfPure} <= v.flags and 
+          def.kind == nkSym and def.sym.kind in {skVar, skLet, skParam} and not 
+          ({sfGlobal, sfPure} <= def.sym.flags or
+            sfCompileTime in def.sym.flags):
+        localError(c.config, def.info, "cannot asign to global variable")
       suggestSym(c.graph, v.info, v, c.graph.usageSym)
 
 proc semConst(c: PContext, n: PNode): PNode =

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -602,7 +602,7 @@ proc isUseLocalVar(n: PNode): bool =
 
 proc globalVarInitCheck(c: PContext, s: PSym, n: PNode) =
   if n.kind in nkCallKinds and isUseLocalVar(n):
-    localError(c.config, n.info, errCannotAsignToGlobal)
+    localError(c.config, n.info, errCannotAssignToGlobal)
   elif n.kind == nkSym and 
       n.sym.kind in {skVar, skLet} and not 
       ({sfGlobal, sfPure} <= n.sym.flags or

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -607,7 +607,7 @@ proc globalVarInitCheck(c: PContext, s: PSym, n: PNode) =
       n.sym.kind in {skVar, skLet} and not 
       ({sfGlobal, sfPure} <= n.sym.flags or
         sfCompileTime in n.sym.flags):
-    localError(c.config, n.info, errCannotAsignToGlobal)
+    localError(c.config, n.info, errCannotAssignToGlobal)
 
 proc semVarOrLet(c: PContext, n: PNode, symkind: TSymKind): PNode =
   var b: PNode

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -36,7 +36,7 @@ const
   errRecursiveDependencyX = "recursive dependency: '$1'"
   errRecursiveDependencyIteratorX = "recursion is not supported in iterators: '$1'"
   errPragmaOnlyInHeaderOfProcX = "pragmas are only allowed in the header of a proc; redefinition of $1"
-  errCannotAssignToGlobal = "cannot assign to global variable"
+  errCannotAssignToGlobal = "cannot assign local to global variable"
 
 proc semDiscard(c: PContext, n: PNode): PNode =
   result = n
@@ -591,22 +591,22 @@ proc msgSymChoiceUseQualifier(c: PContext; n: PNode; note = errGenerated) =
     inc i
   message(c.config, n.info, note, err)
 
-proc isUseLocalVar(n: PNode): bool =
+template isLocalVarSym(n: PNode): bool =
+  n.kind == nkSym and 
+    n.sym.kind in {skVar, skLet} and not 
+    ({sfGlobal, sfPure} <= n.sym.flags or
+      sfCompileTime in n.sym.flags)
+  
+proc usesLocalVar(n: PNode): bool =
   for z in 1 ..< n.len:
-    if n[z].kind == nkSym and n[z].sym.kind in {skVar, skLet} and 
-      not ({sfGlobal, sfPure} <= n[z].sym.flags):
+    if n[z].isLocalVarSym:
       return true
     elif n[z].kind in nkCallKinds:
-      if isUseLocalVar(n[z]):
+      if usesLocalVar(n[z]):
         return true
 
-proc globalVarInitCheck(c: PContext, s: PSym, n: PNode) =
-  if n.kind in nkCallKinds and isUseLocalVar(n):
-    localError(c.config, n.info, errCannotAssignToGlobal)
-  elif n.kind == nkSym and 
-      n.sym.kind in {skVar, skLet} and not 
-      ({sfGlobal, sfPure} <= n.sym.flags or
-        sfCompileTime in n.sym.flags):
+proc globalVarInitCheck(c: PContext, n: PNode) =
+  if n.isLocalVarSym or n.kind in nkCallKinds and usesLocalVar(n):
     localError(c.config, n.info, errCannotAssignToGlobal)
 
 proc semVarOrLet(c: PContext, n: PNode, symkind: TSymKind): PNode =
@@ -761,7 +761,7 @@ proc semVarOrLet(c: PContext, n: PNode, symkind: TSymKind): PNode =
         message(c.config, v.info, hintGlobalVar)
       let isGlobal = {sfGlobal, sfPure} <= v.flags
       if isGlobal:
-        globalVarInitCheck(c, v, def)
+        globalVarInitCheck(c, def)
       suggestSym(c.graph, v.info, v, c.graph.usageSym)
 
 proc semConst(c: PContext, n: PNode): PNode =

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -36,7 +36,7 @@ const
   errRecursiveDependencyX = "recursive dependency: '$1'"
   errRecursiveDependencyIteratorX = "recursion is not supported in iterators: '$1'"
   errPragmaOnlyInHeaderOfProcX = "pragmas are only allowed in the header of a proc; redefinition of $1"
-  errCannotAsignToGlobal = "cannot asign to global variable"
+  errCannotAssignToGlobal = "cannot assign to global variable"
 
 proc semDiscard(c: PContext, n: PNode): PNode =
   result = n

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -759,8 +759,7 @@ proc semVarOrLet(c: PContext, n: PNode, symkind: TSymKind): PNode =
         vm.setupCompileTimeVar(c.module, c.idgen, c.graph, x)
       if v.flags * {sfGlobal, sfThread} == {sfGlobal}:
         message(c.config, v.info, hintGlobalVar)
-      let isGlobal = {sfGlobal, sfPure} <= v.flags
-      if isGlobal:
+      if {sfGlobal, sfPure} <= v.flags:
         globalVarInitCheck(c, def)
       suggestSym(c.graph, v.info, v, c.graph.usageSym)
 

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -596,15 +596,14 @@ proc isUseLocalVar(n: PNode): bool =
     if n[z].kind == nkSym and n[z].sym.kind in {skVar, skLet} and 
       not ({sfGlobal, sfPure} <= n[z].sym.flags):
       return true
-    elif n[z].kind == nkCall:
+    elif n[z].kind in nkCallKinds:
       if isUseLocalVar(n[z]):
         return true
 
 proc globalVarInitCheck(c: PContext, s: PSym, n: PNode) =
-  let isGlobal = {sfGlobal, sfPure} <= s.flags
-  if isGlobal and n.kind == nkCall and isUseLocalVar(n):
+  if n.kind in nkCallKinds and isUseLocalVar(n):
     localError(c.config, n.info, errCannotAsignToGlobal)
-  if isGlobal and n.kind == nkSym and 
+  elif n.kind == nkSym and 
       n.sym.kind in {skVar, skLet} and not 
       ({sfGlobal, sfPure} <= n.sym.flags or
         sfCompileTime in n.sym.flags):
@@ -760,7 +759,9 @@ proc semVarOrLet(c: PContext, n: PNode, symkind: TSymKind): PNode =
         vm.setupCompileTimeVar(c.module, c.idgen, c.graph, x)
       if v.flags * {sfGlobal, sfThread} == {sfGlobal}:
         message(c.config, v.info, hintGlobalVar)
-      globalVarInitCheck(c, v, def)
+      let isGlobal = {sfGlobal, sfPure} <= v.flags
+      if isGlobal:
+        globalVarInitCheck(c, v, def)
       suggestSym(c.graph, v.info, v, c.graph.usageSym)
 
 proc semConst(c: PContext, n: PNode): PNode =

--- a/tests/global/t3505.nim
+++ b/tests/global/t3505.nim
@@ -3,7 +3,7 @@ cmd: "nim check $options --hints:off $file"
 action: "reject"
 nimout: '''
 t3505.nim(22, 22) Error: cannot asign to global variable
-t3505.nim(30, 27) Error: cannot asign to global variable
+t3505.nim(31, 28) Error: cannot asign to global variable
 
 
 
@@ -24,9 +24,10 @@ foo()
 
 # issue #5132
 proc initX(it: float): int = 8
+proc initX2(it: int): int = it
 
 proc main() =
   var f: float
-  var x {.global.} = initX(f)
+  var x {.global.} = initX2(initX(f))
   
 main()

--- a/tests/global/t3505.nim
+++ b/tests/global/t3505.nim
@@ -1,9 +1,32 @@
 discard """
-  errormsg: "cannot asign to global variable"
-  line: 8
+cmd: "nim check $options --hints:off $file"
+action: "reject"
+nimout: '''
+t3505.nim(22, 22) Error: cannot asign to global variable
+t3505.nim(30, 27) Error: cannot asign to global variable
+
+
+
+
+
+'''
 """
+
+
+
+
+
 
 proc foo =
   let a = 0
   var b {.global.} = a
 foo()
+
+# issue #5132
+proc initX(it: float): int = 8
+
+proc main() =
+  var f: float
+  var x {.global.} = initX(f)
+  
+main()

--- a/tests/global/t3505.nim
+++ b/tests/global/t3505.nim
@@ -1,0 +1,9 @@
+discard """
+  errormsg: "cannot asign to global variable"
+  line: 8
+"""
+
+proc foo =
+  let a = 0
+  var b {.global.} = a
+foo()

--- a/tests/global/t3505.nim
+++ b/tests/global/t3505.nim
@@ -2,8 +2,8 @@ discard """
 cmd: "nim check $options --hints:off $file"
 action: "reject"
 nimout: '''
-t3505.nim(22, 22) Error: cannot asign to global variable
-t3505.nim(31, 28) Error: cannot asign to global variable
+t3505.nim(22, 22) Error: cannot assign local to global variable
+t3505.nim(31, 28) Error: cannot assign local to global variable
 
 
 

